File: leetcode/[0003]longest-substring-without-repeating-characters.ts
// https://leetcode.com/problems/longest-substring-without-repeating-characters/
function lengthOfLongestSubstring(s: string): number {
    return solution2(s);
}



// brute force: time limit exceeded
function solution1(s: string): number {
    const sarr = [...new Intl.Segmenter().segment(s)].map((x) => x.segment);

    for (let size = sarr.length; 0 < size; size--) {
        for (let i = 0; i + size - 1 < sarr.length; i++) {
            const sub = sarr.slice(i, i + size);
            if (new Set(sub).size === sub.length) return size;
        }
    }

    return 0;
}



// duplicate char's prev index: 72 ms
function solution2(s: string): number {
    const sarr = [...new Intl.Segmenter().segment(s)].map((x) => x.segment);
    let maxl = 0;

    let prev = 0;
    for (const [i, x] of sarr.entries()) {
        const idx = sarr.slice(prev, i).indexOf(x);
        if (idx !== -1) {
            prev += idx + 1;
        }

        maxl = Math.max(maxl, i - prev + 1);
    }

    return maxl;
}

File: leetcode/[0003]longest-substring-without-repeating-characters.rs
// https://leetcode.com/problems/longest-substring-without-repeating-characters/
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        return solution2(s);
    }
}



// brute force: time limit exceeded
fn solution1(s: String) -> i32 {
    use std::collections::HashSet;

    let s = s.chars().collect::<Vec<char>>();

    for size in (1..=s.len()).rev() {
        for i in 0..=(s.len() - size) {
            let slice = &s[i..(i + size)];
            let set = HashSet::<&char>::from_iter(slice);
            if slice.len() == set.len() {
                return size as i32;
            }
        }
    }

    return 0;
}



// calc prev index: 0 ms
fn solution2(s: String) -> i32 {
    use std::cmp::max;

    let s = s.chars().collect::<Vec<char>>();
    let mut maxl = 0;

    let mut prev_i = 0;
    for (i, &x) in s.iter().enumerate() {
        if let Some(idx) = s[prev_i..i].into_iter().position(|y| *y == x) {
            prev_i += idx + 1;
        }

        maxl = max(maxl, i - prev_i + 1);
    }

    return maxl as i32;
}
